src/record_ts.cpp

```diff
@@ -0,0 +1,1446 @@
+
+                ParamInfo dst_info     = this->dependencies[dependency_index];
+                ReplayVariable &dst_rv = replay_variables[dst_info.slot];
+                VarType vt             = dst_info.vtype;
+                uint32_t tsize         = type_size[(uint32_t) vt];
+                uint32_t workers       = pool_size() + 1;
+
+                uint32_t size;
+                void *src_ptr = 0;
+                if (memcpy) {
+                    ParamInfo src_info =
+                        this->dependencies[dependency_index + 1];
+                    ReplayVariable &src_rv = replay_variables[src_info.slot];
+                    size                   = src_rv.size(vt);
+                    jitc_log(LogLevel::Debug,
+                             "jitc_memcpy_async(dst=%p, src=%p, size=%zu)",
+                             dst_rv.data, src_rv.data, (size_t) size * tsize);
+
+                    src_ptr = src_rv.data;
+                } else {
+                    // Case where in jitc_var_expand, v->is_literal &&
+                    // v->literal == identity
+                    size = op.size;
+                    jitc_log(LogLevel::Debug,
+                             "jitc_memcpy_async(dst=%p, src= literal 0x%lx, "
+                             "size=%zu)",
+                             dst_rv.data, op.data, (size_t) size * tsize);
+                }
+
+                if (size != op.size)
+                    return false;
+
+                uint32_t replication_per_worker =
+                    size == 1u ? (64u / tsize) : 1u;
+                size_t new_size =
+                    size * (size_t) replication_per_worker * (size_t) workers;
+
+                dst_rv.alloc(backend, new_size, dst_info.vtype);
+
+                if (!dry_run)
+                    ts->memset_async(dst_rv.data, new_size, tsize, &op.data);
+
+                if (!dry_run && memcpy)
+                    ts->memcpy_async(dst_rv.data, src_ptr,
+                                     (size_t) size * tsize);
+
+                dst_rv.data_size = size * type_size[(uint32_t) dst_info.vtype];
+                // dst_rv.size = size;
+            };
+            break;
+            case OpType::Compress: {
+                ProfilerPhase profiler(pr_compress);
+
+                uint32_t dependency_index = op.dependency_range.first;
+
+                ParamInfo in_info  = this->dependencies[dependency_index];
+                ParamInfo out_info = this->dependencies[dependency_index + 1];
+
+                ReplayVariable &in_rv  = replay_variables[in_info.slot];
+                ReplayVariable &out_rv = replay_variables[out_info.slot];
+
+                uint32_t size = in_rv.size(in_info.vtype);
+                out_rv.alloc(backend, size, out_info.vtype);
+
+                if (dry_run)
+                    jitc_fail(
+                        "replay(): dry_run compress operation not supported!");
```

Comment:
If the user runs into this error, what could they do about it?
I guess the error message should explain what kind of operations require a dry-run, and say that you can't combine these dry-run ops together with `compress` in the same frozen function?

Reply:
This should be fine, since returning `false` in dry run mode will cause a re-trace of the function. It will however be a bit slower (it has to do the dry run) that's why I put the warning there (I'll expand on it in the warning message).

Does this make sense?
```cpp
                    jitc_log(LogLevel::Warn,
                             "replay(): Could not infer the size of the result "
                             "to a dr.compress call during dry run. The "
                             "function will be re-recorded, which causes some "
                             "overhead. This occured most likely due to a "
                             "scatter_reduce call in a LLVM kernel in the same "
                             "frozen function. To avoid this try to split the "
                             "function into two frozen functions.");

```